### ISSUE TITLE
SOLVED: Trouble enabling the JSON API functionality

### ISSUE THREAD

**Issue by topcat** [RetroShare/RetroShare#2179]
Hello,

I have just cloned and built RS from master and built it with the following qmake options :

qmake RS_UPNP_LIB=miniupnpc CONFIG+="rs_jsonapi rs_webui"

Basically I have tried to following instructions at :

https://github.com/RetroShare/RetroShare/issues/2070
https://github.com/retroshare/rsnewwebui

As per the above instructions I should see under preferences a tab for JSON API and I then need to enable it before Retroshare-service can accept JSON API requests. Please forgive me if I have misunderstood this.

I do not see the JSON API tab under preferences.

I am building RS on Debian Unstable with the rapidjson-dev package ver 1.1.0+dfsg2-6 installed. The build seems to proceed without any errors. 

Port 9092, which seems to be the default to handle JSON requests, is not being listened on, neither when I use the RS GUI nor if I run the retroshare-service.

Apologies if I am missing something obvious that has been covered here before. I would appreciate any help on how to enable JSON API functionality.

Thank you.

**Comment by sehraf** [comment]
can you try `qmake CONFIG+=rs_jsonapi CONFIG+=rs_webui 'RS_UPNP_LIB=miniupnpc' ` ?

**Comment by topcat** [comment]
Hi sehraf, thanks for replying back. I just tried what you suggested but no luck. Still don't see the JSON API section under Preferences. Please let me know if I should try something else or if you need more data. Thanks again :)

**Comment by defnax** [comment]
other option i get often issues on windows to enable with qmake the CONFIG+

i edit then my **retroshare.pri**
```
# To enable JSON API append the following assignation to qmake command line
# "CONFIG+=rs_jsonapi"
CONFIG+=rs_jsonapi
```
```
# To enable webui append the following assignation to qmake
# command line "CONFIG+=rs_webui"
CONFIG+=rs_webui
```

**Comment by sehraf** [comment]
Does qmake complain about anything missing? Did you do a full rebuild? You should see a lot of doxygen output and maaaany lines of output from the python script when generating the API.

Can you also post the content of the folder `jsonapi-generator/src` ? Should look like this: 
`jsonapi-generator  jsonapi-generator-doxygen-final.conf  jsonapi-generator.o  jsonapi-includes.inl  jsonapi-wrappers.inl  Makefile.jsonapi-generator  xml` with many xml files in the `xml` folder

**Comment by topcat** [comment]
Thank you defnax and sehraf.

Sehraf, the contents of the jsonapi-generator/src file shows the following
files :

async-method-wrapper-template.cpp.tmpl
jsonapi-generator
jsonapi-generator.cpp
jsonapi-generator-doxygen.conf
jsonapi-generator.o
jsonapi-generator.pro
Makefile.jsonapi-generator
method-wrapper-template.cpp.tmpl

There doesn't exist a xml folder and no I didn't see a lot of doxygen
output.

If I do a "make clean" and then run qmake again with the options, the
following is the output I see :

Project MESSAGE: RetroShare version 0.6.6-RC2-26-ge367a1f69 determined via
git
Project WARNING: QMAKE: You have disabled deprecated warnings.
Project WARNING: QMAKE: You have disabled C preprocessor warnings.
Project WARNING: QMAKE: You have enabled RetroShare direct chat which is
deprecated!
Project MESSAGE: using system rapidjson
Must specify package names on the command line


Maybe the rs_jsonapi isn't getting passed through. I will try what defnax
has suggested too. Please let me know if you would like me to try something
else as well. Thank you.

**Comment by sehraf** [comment]
> Must specify package names on the command line

looks fishy, but lets see what defnax suggestion does.

**Comment by defnax** [comment]
possible my method doesnt helps, i wasnt able on windows to compile rs with jason api clean.

**Comment by topcat** [comment]
Hello sehraf / defnax, 

Problem solved :)

Since the jsonapi-generator xml files were not being generated even after doing "make clean" and doing a git pull to the latest master followed by multiple build attempts, I decided to blow my retroshare directory and start with an absolutely fresh clone. 

I used the same qmake commands as sehraf has pointed to me earlier :

qmake CONFIG+=rs_jsonapi CONFIG+=rs_webui 'RS_UPNP_LIB=miniupnpc'

This seems to have helped since now I have the xml folder with a ton of xml files and retroshare-gui now has the JSON API tab under Preferences :)

The build process this time around was markedly different from earlier attempts. With a fresh clone, I ran into an early issue trying to clone restbed under supportlibs/restbed. It failed trying to do a clone from remote with the message that the directory was not empty. I saw there was an empty include folder and deleted that manually. Thereafter the clone and subsequent build of restbed went ahead smoothly.

I then ran into an issue where I was seeing a lot of undefined reference errors to restbed during linking. I realised that the linker was using the restbed already installed on my Debian system via the librestbed-dev and librestbed0 packages. I removed these and then the linker linked against the librestbed.a library generated by the retroshare build process and it's all good now :)

Thank you so much sehraf / defnax for your help and support through this process :) Appreciate the help.
